Os/Linux/SystemResources.cpp

```diff
@@ -0,0 +1,137 @@
+// \title  Linux/SystemResources.cpp
+// \author sfregoso
+// \brief  hpp file for SystemResources component implementation class
+//
+// \copyright
+// Copyright 2021, by the California Institute of Technology.
+// ALL RIGHTS RESERVED.  United States Government Sponsorship
+// acknowledged.
+//
+// ======================================================================
+#include <cstdio>              /* fopen() */
+#include <cstdlib>             /* scanf */
+#include <sys/vfs.h>            /* statfs() */
+#include <cstring>
+#include <Os/SystemResources.hpp>
+#include <Fw/Types/Assert.hpp>
+
+namespace Os {
+    static const U32 LINUX_CPU_LINE_LIMIT = 1024; // Maximum lines to read before bailing
+
+    SystemResources::SystemResourcesStatus SystemResources::getCpuCount(U32 &cpuCount) {
+        char line[512] = {0};
+        FILE *fp = nullptr;
+        U32 cpu_count = 0;
+
+        if ((fp = fopen("/proc/stat", "r")) == nullptr) {
+            return SYSTEM_RESOURCES_ERROR;
+        }
+
+        if (fgets(line, sizeof(line), fp) == nullptr) { //1st line.  Aggregate cpu line. Skip
+            fclose(fp);
+            return SYSTEM_RESOURCES_ERROR;
+        }
+
+        for (U32 i = 0; i < LINUX_CPU_LINE_LIMIT; i++) {
+            if (fgets(line, sizeof(line), fp) == nullptr) { //cpu# line
+                break;
+            }
+
+            if (!(line[0] == 'c' && line[1] == 'p' && line[2] == 'u')) {
+                break;
+            }
+            cpu_count++;
+        }
+        fclose(fp);
+        cpuCount = cpu_count;
+
+        return SYSTEM_RESOURCES_OK;
+    }
+
+    SystemResources::SystemResourcesStatus SystemResources::getCpuTicks(CpuTicks &cpu_ticks, U32 cpu_index) {
+        char line[512] = {0};
+        FILE *fp = nullptr;
+        U32 cpu_data[4] = {0};
+        U32 cpuCount = 0;
+        SystemResources::SystemResourcesStatus status  = SYSTEM_RESOURCES_ERROR;
+        U64 cpuUsed = 0;
+        U64 cpuTotal = 0;
+
+        if ((status = getCpuCount(cpuCount)) != SYSTEM_RESOURCES_OK) {
+            return status;
+        }
+
+        if (cpu_index >= cpuCount) {
+            return SYSTEM_RESOURCES_ERROR;
+        }
+
+        if ((fp = fopen("/proc/stat", "r")) == nullptr) {
+
+            return SYSTEM_RESOURCES_ERROR;
+
+        }
+
+        if (fgets(line, sizeof(line), fp) == nullptr) { //1st line.  Aggregate cpu line.
+            fclose(fp);
+            return SYSTEM_RESOURCES_ERROR;
+        }
+
+        for (U32 i = 0; i < cpu_index + 1; i++) {
+            if (fgets(line, sizeof(line), fp) == nullptr) { //cpu# line
+                fclose(fp);
+                return SYSTEM_RESOURCES_ERROR;
+            }
+            if (i != cpu_index) { continue; }
+
+            if (!(line[0] == 'c' && line[1] == 'p' && line[2] == 'u')) {
+                fclose(fp);
+                return SYSTEM_RESOURCES_ERROR;
+            }
+
+            sscanf(line, "%*s %d %d %d %d", &cpu_data[0],
```

Comment:
The uses of sscanf and fscanf make me a bit nervous. I think they are OK, because no strings are stored, only numbers. Strings are scanned but discarded. This operation should be memory safe. Maybe say that in a comment.